apps/glusterfs/device_entry.go

```diff
@@ -243,12 +260,23 @@ func (d *DeviceEntry) SetState(tx *bolt.Tx,
 			if err != nil {
 				return err
 			}
+			d.State = s
+			err = d.Save(tx)
+			if err != nil {
+				return err
+			}
 		case api.EntryStateFailed:
-			// Only thing to do here is to set the state
+
+			err := d.Remove(tx, e, a)
```

Comment:
This is the most important call.

It looks like this function could instead be passed `db` like volume.Create(), and let it pick, when and when not to stay in context.